Fix Reject ACLs and improve logging

trozet opened the pull request:
Fix Reject ACLs for NodePort and External IP, and node add
    
    The previous patch for moving Reject ACLs to port group only applies the
    ACLs to the port group with management ports, which means the ACL will
    only get applied to the node logical switch. For NodePort and External
    IP those ACLs need to be on the External switches of the GR where the
    load balancer is hosted.
    
    Also, the reject ACLs were not getting created/deleted correctly and
    some of the unit tests were missing the expected commands.

   During node add, we were still adding reject ACLs directly to switches,
   instead of letting them naturally take effect through the mgmt port
   being added to the port group.
    
    Signed-off-by: Tim Rozet <trozet@redhat.com>

trozet commented:
looks like a race problem: [Fail] Informer Event Handler Tests [It] adds existing pod and processes an update event 
/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/informer/informer_test.go:267

@aojea @alexanderConstantinescu

trozet commented:
/retest

trozet commented:
@abhat @vishnoianil @dcbw

vishnoianil commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -269,15 +293,27 @@ func (ovn *Controller) deleteLoadBalancerRejectACL(lb, vip string) {
-		if err == nil {
-			ovn.removeACLFromPortGroup(lb, aclUUID)
+		if err != nil {
+			klog.Infof("Unable to delete Reject ACL for load-balancer: %s, vip: %s. No entry in cache and "+
```
log message needs update?

trozet replied:
looks OK to me

vishnoianil replied:
It's trying to findStaleReject, not deleting it, so probably  "Unable to find stale reject ACL....." would be better ?

vishnoianil commented:
> ```
> Fix Reject ACLs for NodePort and External IP
> 
> The previous patch for moving Reject ACLs to port group only applies the
> ACLs to the port group with management ports, which means the ACL will
> only get applied to the node logical switch. For NodePort and External
> IP those ACLs need to be on the External switches of the GR where the
> load balancer is hosted.
It seems like getLogicalSwitchesForLoadBalancer does return the GR external switch, but the management port of this external logical switch is not added to the cluster port-group. So wondering if adding the management port of this external switch to cluster port-group will work, rather than specifically associating the ACL to this external logical-switch?
> 
> Also, the reject ACLs were not getting created/deleted correctly and
> some of the unit tests were missing the expected commands.
So some of the unit tests were incorrectly passing even the expected commands were missing?
> 
> Signed-off-by: Tim Rozet <trozet@redhat.com>
> ```

alexanderConstantinescu commented:
> looks like a race problem: [Fail] Informer Event Handler Tests [It] adds existing pod and processes an update event
> /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/informer/informer_test.go:267
> 
> @aojea @alexanderConstantinescu

I've never hit that. Could you paste more logs next time you see that?

vishnoianil commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -222,11 +233,19 @@ func (ovn *Controller) createLoadBalancerRejectACL(lb, sourceIP string, sourcePo
+			cmd = append(cmd, "--", "add", "port_group", ovn.clusterPortGroupUUID, "acls", aclUUID)
+		}
+		if len(gwRouterExtSwitch) > 0 {
+			cmd = append(cmd, "--", "add", "logical_switch", gwRouterExtSwitch, "acls", aclUUID)
```
it might work, but it will end up installing lot of flows on the external switch as well for normal service reject acl and that the last thing we want. Assuming that reject acl for NodePort/External Ip will be installed on specific external switch of specific node, this makes sense to me.

trozet replied:
yeah management port doesn't go on the external switch, only node switch

vishnoianil commented:
> > ```
> > Fix Reject ACLs for NodePort and External IP
> > 
> > The previous patch for moving Reject ACLs to port group only applies the
> > ACLs to the port group with management ports, which means the ACL will
> > only get applied to the node logical switch. For NodePort and External
> > IP those ACLs need to be on the External switches of the GR where the
> > load balancer is hosted.
> > ```
> 
> It seems like getLogicalSwitchesForLoadBalancer does return the GR external switch, but the management port of this external logical switch is not added to the cluster port-group. So wondering if adding the management port of this external switch to cluster port-group will work, rather than specifically associating the ACL to this external logical-switch?

I think it won't work the way i wanted, it will end up installing all the reject acl flows for services to the external switch as well. Given that NodePort/External Ip reject acl's are installed on specific external logical switch of the nodes, this makes sense to me.

> 
> > Also, the reject ACLs were not getting created/deleted correctly and
> > some of the unit tests were missing the expected commands.
> > So some of the unit tests were incorrectly passing even the expected commands were missing?
> > Signed-off-by: Tim Rozet [trozet@redhat.com](mailto:trozet@redhat.com)

trozet commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -189,16 +191,25 @@ func (ovn *Controller) generateACLNameForOVNCommand(lb string, sourceIP string,
+	if len(switches) > 0 {
+		applyToPortGroup = true
+	} else {
+		klog.V(5).Infof("Ignoring creating reject ACL for port group with load balancer %s. It has no "+
```
It wont create the reject ACL on the port group, but it might still create it on the external switch

vishnoianil replied:
yeah, but condition is checking for number of switches for lb ?

trozet commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -189,16 +191,25 @@ func (ovn *Controller) generateACLNameForOVNCommand(lb string, sourceIP string,
+	}
+
+	// check if the load balancer is on a GR, if so we need to get the external switches
+	gwRouterExtSwitch, err := ovn.getGRLogicalSwitchForLoadBalancer(lb)
```
some load balancers are not on the routers, so we cant just assume if the lb is on the node switch that it is also on the external switch

vishnoianil replied:
@trozet not sure i got it. So is it possible that LB can be on both node switch and GR external switch?

trozet commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -242,10 +261,15 @@ func (ovn *Controller) createLoadBalancerRejectACL(lb, sourceIP string, sourcePo
+	if applyToPortGroup {
+		cmd = append(cmd, "--", "add", "port_group", ovn.clusterPortGroupUUID, "acls", "@reject-acl")
+	}
+	if len(gwRouterExtSwitch) > 0 {
```
they are not mutually exclusive. A reject acl may need to be on both the external switch as well as the node logical switch

vishnoianil replied:
like for single NodePort service, we have to install reject ACL on both node switch and GR logical switch?

trozet commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -242,10 +261,15 @@ func (ovn *Controller) createLoadBalancerRejectACL(lb, sourceIP string, sourcePo
 	cmd := []string{"--id=@reject-acl", "create", "acl", "direction=from-lport", "priority=1000", aclMatch, "action=reject",
 		fmt.Sprintf("name=%s", aclName)}
-	cmd = append(cmd, "--", "add", "port_group", ovn.clusterPortGroupUUID, "acls", "@reject-acl")
+	if applyToPortGroup {
```
yeah if I push another patch I'll clean this up

trozet replied:
will clean this up as I'm going to add another commit for another fix with adding the reject acls accidentally to swtiches

trozet replied:
actually applyToPortGroup is used in multiple places, so just going to leave it as is instead of changing to len(switches) everywhere

trozet commented on go-controller/pkg/ovn/endpoints_test.go:
```diff
@@ -123,6 +123,10 @@ func (e endpoints) delCmds(fexec *ovntest.FakeExec, service v1.Service, endpoint
 			fexec.AddFakeCmdsNoOutputNoError([]string{
 				fmt.Sprintf("ovn-nbctl --timeout=15 --data=bare --no-heading --columns=_uuid find logical_switch load_balancer{>=}%s", k8sTCPLoadBalancerIP),
 				fmt.Sprintf("ovn-nbctl --timeout=15 --data=bare --no-heading --columns=name find logical_router load_balancer{>=}%s", k8sTCPLoadBalancerIP),
+				fmt.Sprintf("ovn-nbctl --timeout=15 --data=bare --no-heading --columns=_uuid find acl name=%s-%s\\:%v",
```
yeah if we delete the endpoints, then a reject ACL should be created for the service

vishnoianil replied:
okay, so the unit test was broken but passing?

trozet replied:
Yes I believe so. Looking back to even 4.5 I'm not sure how this test case was passing, other than there must have been a bug without logic in creating the reject acls:

https://github.com/openshift/ovn-kubernetes/blob/release-4.5/go-controller/pkg/ovn/endpoints_test.go#L202

If we have a service with one endpoint and we delete that endpoint, then a service reject acl should get created.

danwinship replied:
So actually, this looks consistent; the new set of matched commands is consistent with there being no logical switches associated with the load balancer (note that the `create acl` command here isn't followed by an `add logical_switch ... acl` command). In the old version of `createLoadBalancerRejectACL` we would return early without creating the ACL in that case, while in the new version, we create the ACL even if there's currently nothing using it.

although... actually, can you do that? OVS is pretty aggressive about garbage-collecting unused objects. Can you create an ACL but not assign it to anything? (Or does this case never get hit outside of the unit tests anyway?)

trozet replied:
@danwinship nice catch! you are right in this unusual case where we have no switches at all, we should not apply the reject acl to anything. This isn't exactly useful for our unit test other than catching this. So perhaps I should also change the service_test.go to actually have some switches with the load balancer on them. I'll fix the code to not create the acl if no switches or port groups to apply to.

trozet commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -269,15 +293,27 @@ func (ovn *Controller) deleteLoadBalancerRejectACL(lb, vip string) {
+	gwRouterSwitch, err := ovn.getGRLogicalSwitchForLoadBalancer(lb)
+	if err != nil {
+		klog.Errorf("Unable to query logical switches for GR with load balancer: %s, error: %v", lb, err)
+	} else {
```
removeACLFromNodeSwitches handles if it is an empty slice

trozet commented:
/retest

trozet commented:
/hold while I test this downstream

trozet commented:
/hold cancel
this is now passing downstream https://github.com/openshift/ovn-kubernetes/pull/331

trozet commented:
@girishmg

danwinship commented on go-controller/pkg/ovn/service.go:
```diff
@@ -169,14 +171,33 @@ func (ovn *Controller) syncServices(services []interface{}) {
 							klog.Infof("Service Sync: Removing OVN stale reject ACL: %s", name)
 							ovn.removeACLFromPortGroup(lb, uuid)
+							var foundSwitches []string
+							// For upgrade from a non-port group Reject ACL implementation
```
should there be a note that we can eventually remove this code, like there used to be on `removeACLFromNodeSwitches`?

trozet replied:
yeah I can add a comment here. Note, I removed that comment from the removeACLFromNodeSwitches as we need it for ACLs on the external switches now and in the foreseeable future.

danwinship reviewed (commented):
not super-familiar with the code, but if it fixes the tests then yay

trozet commented on go-controller/pkg/ovn/service.go:
```diff
@@ -332,7 +332,8 @@ func (ovn *Controller) createService(service *kapi.Service) error {
 						}
-						klog.Infof("Service Reject ACL created for gateway router: %s", aclUUID)
+						klog.Infof("Service Reject ACL created for NodePort service: %s, via gateway "+
+							"router: %s:%s:%d, ACL UUID:%s", service.Name, svcPort.Protocol, physicalIP, port, aclUUID)
```
will do

trozet commented:
/retest

danwinship commented:
looks good other than the worry about creating unused ACLs

trozet commented:
ok @danwinship  take a look now. I added a statement to not create the ACL if there are no switches for the load balancer. Changed the endpoints_test to check this condition. Modified the service_test to actually simulate that there is an switch for the load balancer, and ensure the command applies the ACL to the port group.

trozet commented:
egress IP flake:
https://github.com/ovn-org/ovn-kubernetes/pull/1785/checks?check_run_id=1364372179

trozet commented:
/retest

trozet commented:
/retest

danwinship commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -190,16 +194,30 @@ func (ovn *Controller) generateACLNameForOVNCommand(lb string, sourceIP string,
+
+	if len(switches) == 0 && gwRouterExtSwitch == "" {
+		return "", fmt.Errorf("load balancer %s does not apply to any switches in the cluster. Will not create "+
+			"Reject ACL", lb)
```
Can this actually happen? It looks like this error will propagate up and result in the service being half-added.

trozet replied:
It's extremely unlikely, but it's not bad to add some code to catch this case. Let's break down what happens for each possible createLoadBalancerRejectACL call:
1. endpoints deletion: if all the endpoints are gone for a service, try to create a reject ACL. If no switches, create no ACL. Load balancer endpoints are still cleared. The result will just be no reject happens when a client tries to access the LB. But that LB is not on any network entity so it is a moot point. Behavior here is OK.
2. service creation: if this errors because the load balancer is not on a switch, then yeah we will end up returning out of the service add. So the other potential svc Ports (if using a different protocol) may not get their reject ACL created. This could be a pitfall if somehow one protocol load balancer was on a switch but another wasn't. If you want I can modify the service add code to handle this case. Let me know what you think.

danwinship commented:
lgtm

danwinship commented:
/retest
